Accept multiple ingest pipelines in Filebeat

ycombinator opened the pull request:
Motivated by https://github.com/elastic/beats/pull/8852#issuecomment-434973388.

Starting with 6.5.0, Elasticsearch Ingest Pipelines have gained the ability to:
- run sub-pipelines via the [`pipeline` processor](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/pipeline-processor.html), and
- conditionally run processors via an [`if` field](https://www.elastic.co/guide/en/elasticsearch/reference/6.5/ingest-processors.html).

These abilities combined present the opportunity for a fileset to ingest the same _logical_ information presented in different formats, e.g. plaintext vs. json versions of the same log files. Imagine an entry point ingest pipeline that detects the format of a log entry and then conditionally delegates further processing of that log entry, depending on the format, to another pipeline.

This PR allows filesets to specify one or more ingest pipelines via the `ingest_pipeline` property in their `manifest.yml`. If more than one ingest pipeline is specified, the first one is taken to be the entry point ingest pipeline.

#### Example with multiple pipelines
```yaml
ingest_pipeline:
  - pipeline-ze-boss.json 
  - pipeline-plain.json
  - pipeline-json.json
```
#### Example with a single pipeline
_This is just to show that the existing functionality will continue to work as-is._
```yaml
ingest_pipeline: pipeline.json
```

Now, if the root pipeline wants to delegate processing to another pipeline, it must use a `pipeline` processor to do so. This processor's `name` field will need to reference the other pipeline by its name. To ensure correct referencing, the `name` field must be specified as follows:

```json
{
  "pipeline" : {
    "name": "{< IngestPipeline "pipeline-plain" >}"
  }
}
```

This will ensure that the specified name gets correctly converted to the corresponding name in Elasticsearch, since Filebeat prefixes it's "raw" Ingest pipeline names with `filebeat-<version>-<module>-<fileset>-` when loading them into Elasticsearch.

ycombinator commented:
There's still one TODO item left for this PR related to documentation but I think it's ready for a code review. Thanks!

ycombinator commented:
There's still one TODO item left for this PR related to documentation but I think it's ready for a code review. Thanks folks!

ycombinator commented on CHANGELOG-developer.asciidoc:
```diff
@@ -64,3 +64,4 @@ The list below covers the major changes between 6.3.0 and master only.
 - Allow to disable config resolver using the `Settings.DisableConfigResolver` field when initializing libbeat. {pull}8769[8769]
 - Add `mage.AddPlatforms` to allow to specify dependent platforms when building a beat. {pull}8889[8889]
 - Add `cfgwarn.CheckRemoved6xSetting(s)` to display a warning for options removed in 7.0. {pull}8909[8909]
+- Filesets can now define multiple ingest pipelines, with the first one considered as the root pipeline. {pull}8914[8914]
```
Yes, I'm happy to change the terminology to whatever is clearest to most folks.

ycombinator commented:
jenkins, test this

ycombinator commented:
@dedemorton I've added docs for this feature in df2488a27e171a8f10d8fac6766ae7a352dbf1a8 but I'm not sure about my language and structure/organization. I'd love for you to take a look, if you have some time. Thanks!

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ruflin commented:
What happens if a user uses a module with multiple pipelines against and older version of Elasticsearch?

ycombinator commented:
> What happens if a user uses a module with multiple pipelines against and older version of Elasticsearch?

I tested this PR with Elasticsearch 6.4.0, where neither the `pipelines` processor nor the `if` field on processors exist.

There are two related parts to this PR. One worked with ES 6.4.0, but the other did not. Specifically:

1. The part where a module developer can now specify multiple ingest pipelines under the `ingest_pipeline` key in the module's `manifest.yml` file will work with older versions of ES. All pipelines listed will get loaded into ES.

2. The part where a pipeline uses the `pipeline` processor or the `if` processor field, both features introduced in ES 6.5 will not work with older versions of ES. Concretely, when Filebeat attempts to load such a pipeline into ES, ES will return a 400 with an error message that Filebeat emits to its log. FWIW, such an error can be returned today (in `master` as well, if there is a processor in a pipeline that has not been installed in ES as a plugin, e.g. the CSV processor plugin).

   Note that the pipeline loading (within a fileset) is short-circuiting. For example, imagine that a module developer has specified 3 pipelines in the list under `ingest_pipeline` in the module's `manifest.yml`. Let's say the 2nd one in the list fails loading because it contains a `pipeline` processor or an `if` somewhere. The 1st pipeline in the list would've been successfully loaded into ES but the 3rd one will not get loaded. And obviously the 2nd one won't get loaded either because it caused the ES error.

   I can see four options on how to proceed:

   1. Don't load the pipelines (within a fileset) in a short-circuiting fashion. Instead, even if one pipeline fails to load for some reason, emit the error in the log, but continue to try and load the remaining pipelines in the list.
   
   2. If a pipeline in the list fails to load, try to `DELETE` previously-loaded pipelines for the fileset. This way either all pipelines or a fileset get loaded or none of them do.

   3. Before attempting to load _any_ pipelines for a fileset, ask Elasticsearch to validate _all_ of them. Again, this way either all pipelines or a fileset get loaded or none of them do.

   4. Keep it simple and keep the current short-circuiting approach. Depending on the error, users might need to upgrade ES or perform some other fix, such as loading a plugin, or filing a bug to get the pipeline definition in Filebeat fixed up. Once the fix is made, they can simply re-run Filebeat and any previously-erroring pipelines will get loaded into ES.

   My personal preference would be option 1. It is essentially like option 4 but doesn't give up on the remaining pipelines either, so if multiple pipelines have different errors, at least the user can find out about them all at once. Option 2 could lead to an inconsistent state (if any of the `DELETE`s fail). Option 3 is probably the safest but also unjustifiably expensive.

kvch commented:
I would not go with option 1, because it pollutes the ES instance of users with possibly unused pipelines.

ruflin commented:
Can you elaborate on why you think 3 is too expensive? I would expect checking for the pipelines does not happen too often so I would not be worried about the extra load.

I just realised this also touches and other problem: What if geo or user_agent are not installed? At the moment I think it does not get loaded as we only have one pipeline but not sure how good the error is. We already have requirements for the pipelines in the manifest, there we could also add requirements for the ES version and use option 3?

ycombinator commented:
> Can you elaborate on why you think 3 is too expensive? I would expect checking for the pipelines does not happen too often so I would not be worried about the extra load.

It might not be. It would depend on how many multi-pipeline filesets we end up with over time. For every such fileset, we would call an ES API (the simulate ingest pipeline API) for each pipeline in that fileset. As you note, all of this would only happen at startup time. So yeah, maybe compared to the overall runtime of the Filebeat process it's not too expensive, just makes the initial startup time a bit longer.

kvch commented:
I am fine with that option. Just keep in mind that simulating multiple pipelines are not yet implemented: https://github.com/elastic/elasticsearch/issues/35495
Also, I have implemented simulating pipelines in a script: https://github.com/elastic/beats/blob/master/filebeat/scripts/tester/main.go#L252 You might find something useful there.

ycombinator commented:
Ah, great point about not being able to simulate pipelines referencing other pipelines without creating the "inner" pipelines first!

I do think that puts a dampener on option 3 (pre-validation) for now, until elastic/elasticsearch#35495 is resolved. It doesn't make sense to create the "inner" pipelines (assuming they pass simulation), then try to simulate the "outer" pipeline only to have it fail. At that point, we might as well try to push them all and then rollback whatever we created upon failure (option 2).

So now I'm in favor of option 2 (rollbacks) 😄 @ruflin: your thoughts now?

ruflin commented:
What are the reasons loading pipelines could fail besides wrong Elasticsearch version or missing plugins?

What if Elasticsearch would actually support to bulk load related pipelines? I'm kind of thinking it should and then provide us with the all or nothing feature?

ycombinator commented:
> What if Elasticsearch would actually support to bulk load related pipelines? I'm kind of thinking it should and then provide us with the all or nothing feature?

Created https://github.com/elastic/elasticsearch/issues/35889 to request this feature.

ycombinator commented:
@ruflin I chatted with @jakelandis and @jaymode off-PR about the two ES feature requests (https://github.com/elastic/elasticsearch/issues/35495, https://github.com/elastic/elasticsearch/issues/35889) that would help implement this PR in a really nice way.

We agreed that both feature requests are valid and good additions, but neither one is likely to be implemented for 6.6.0. So, rather than holding up this PR, I propose we implement the rollback approach for now. As the two ES feature requests get done, we'll change the Filebeat implementation to take advantage of them. The advantage of this is that it would let users ingest Elasticsearch structured audit logs and we can deprecate the unstructured one right away. 

WDYT?

ruflin commented:
SGTM. The part I mainly worry is the complexity around the rollback that we potentially have to keep for quite some time.

An other question we need to resolve as this is targeted for 6.6. Do you also plan to modify some modules on it? If yes, is it a breaking change? Assuming in FB 6.3-6.5 a module was compatible with ES 6.0-6.x, now in FB 6.6 it is only compatible with ES 6.5-6.x Do we consider this a breaking change?

dedemorton commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -336,6 +354,88 @@ Note that you should follow the convention of naming of fields prefixed with the
 `nginx.access.remote_ip`. Also, please review our <<event-conventions>>.
 
+In 6.6 and later, ingest pipelines can use the `pipeline` processor to delegate
+part of the processing to another sub-pipeline.
```
Or maybe so that it doesn't sound singular: "...parts of the processing to other pipelines"

ycombinator commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -229,6 +229,24 @@ This example selects the ingest pipeline file based on the value of the
+  - ingest/json_logs.json
+----
+
+When multiple ingest pipelines are specified the first one in the list is
```
It's possible for the second pipeline in the list to delegate to another pipeline.

However, if a fileset is going to define multiple ingest pipelines, it needs to designate exactly one of them as the entry point. I want to document the idea that fileset authors should place their entry point pipeline in the first position in the list; that's how the code knows which of the listed pipelines is the entry point one.

ycombinator commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -229,6 +229,24 @@ This example selects the ingest pipeline file based on the value of the
+
+One reason for using multiple pipelines might be to send all logs harvested
+by this fileset to the root pipeline and have it delegate different parts of
+the processing to different sub-pipelines. You can read details about setting
```
How about:

> One reason for using multiple pipelines might be to send all logs harvested
by this fileset to the entry point pipeline and have it delegate different parts of
the processing to other pipelines.

dedemorton replied:
sounds OK to me.

ycombinator commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -336,6 +354,88 @@ Note that you should follow the convention of naming of fields prefixed with the
+
+This can be useful if you want a fileset to ingest the same _logical_ information
+presented in different formats, e.g. csv vs. json versions of the same log files.
+Imagine a root ingest pipeline that detects the format of a log entry and then conditionally
```
We are already introducing the "entry point pipeline" terminology further up in this doc. So could we say the following here?

> Imagine an entry point ingest pipeline that detects the format of a log entry and then conditionally delegates further processing of that log entry, depending on the format, to other
pipelines.

dedemorton replied:
Since you describe the behavior of the pipeline, you can just say "ingest pipeline" here.

IMO, fewer words = easier to parse = quicker to read = happier user.

I don't object strongly, though, so please proceed as you see fit.

ycombinator commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -336,6 +354,88 @@ Note that you should follow the convention of naming of fields prefixed with the
+  - ingest/json_logs.json
+----
+
+Additionally, any `pipeline` processors referencing one of the pipelines in the
```
Agreed that the conflicting examples cause confusion. Technically both examples are valid, but the one below, i.e. the one using the `< IngestPipeline "plain_logs" >` syntax, is the one we'd want fileset authors to use when trying to use this new feature (multiple pipelines with an entry point pipeline for a fileset).

The reason I split up the two examples was to first just get the reader introduced to the idea of a pipeline that delegates part of the processing to other pipelines and then explain the idea of the template function. I was wary of introducing **two** new ideas in the same example snippet in case.

But I see now the confusion this can cause, so I agree that we should always show the use of the `< IngestPipeline "plain_logs" >` syntax in the example(s). Perhaps there's some way we could smoosh the examples and explanation of these two ideas together without potentially overwhelming the reader?

ycombinator replied:
Perfect, callouts make a lot of sense here, thanks!

dedemorton commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -336,6 +354,88 @@ Note that you should follow the convention of naming of fields prefixed with the
+{ref}/conditionals-with-multiple-pipelines.html[In 6.6 and later], ingest pipelines can use the
+`pipeline` processor to delegate parts of the processings to other pipelines.
+
+[source,json]
```
I'd remove this snippet.

dedemorton commented on docs/devguide/modules-dev-guide.asciidoc:
```diff
@@ -336,6 +354,88 @@ Note that you should follow the convention of naming of fields prefixed with the
+  - ingest/json_logs.json
+----
+
+Additionally, any `pipeline` processors referencing one of the pipelines in the
```
Then remove this sentence because you shouldn't need it because you show the correct syntax in the full example.

ycombinator commented:
@ruflin @kvch I've implemented the rollback-on-error behavior in e4ceda2. This PR is ready for your 👀 again, when you get a chance. Thanks!

@dedemorton I believe I've incorporated all your excellent doc feedback now. Ready for your 👀 as well, when you get a chance. Thanks as well!

ycombinator commented:
@ruflin said:

> An other question we need to resolve as this is targeted for 6.6. Do you also plan to modify some modules on it? If yes, is it a breaking change? Assuming in FB 6.3-6.5 a module was compatible with ES 6.0-6.x, now in FB 6.6 it is only compatible with ES 6.5-6.x Do we consider this a breaking change?

It's a good question. I don't consider this PR by itself a breaking change since it only introduces the framework for multiple pipelines in a fileset. No modules are changed in this PR. 

However, I think we will need to answer your question as part of the #8852.

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
@urso @ruflin I could use your guidance on whether to merge this PR now or not. Travis CI is green but Jenkins CI is not. The failure in Jenkins CI is because of the `Too many dynamic script compilations within, max: [75/5m]` error which will hopefully be fixed by https://github.com/elastic/beats/pull/9777.

ruflin commented:
@ycombinator I'm good with merging. But in case #9777 gets in before this one, a rebase would be nice.
